darwin/Platform.c

```diff
@@ -725,3 +725,43 @@ void Platform_gettime_monotonic(uint64_t* msec) {
 #endif
 
 }
+
+#ifndef OSRELEASEFILE
+#define OSRELEASEFILE "/System/Library/CoreServices/SystemVersion.plist"
+#endif
+
+static void Platform_getOSRelease(char* buffer, size_t bufferLen) {
+   FILE* fp = fopen(OSRELEASEFILE, "r");
+   if (!fp) {
+      xSnprintf(buffer, bufferLen, "No OS Release");
+      return;
+   }
+
+   /* extract version in: <key>ProductVersion</key>\n<string>X.Y.Z</string>\n */
+   bool getKeyValue = false;
+   char version[64] = {'\0'};
+   char lineBuffer[256];
+   char* sp;
+   while ((sp = fgets(lineBuffer, sizeof(lineBuffer), fp))) {
+      while (*sp != '\0' && isspace(*sp)) sp++;
+      if (String_startsWith(sp, "<key>ProductVersion</key>")) {
+         getKeyValue = true;
+         continue;
+      }
+      if (getKeyValue == false)
+         continue;
+      if (String_startsWith(sp, "<string>")) {
+         const char* start = sp + strlen("<string>");
+         const char* stop = strstr(sp, "</string>");
+         String_safeStrncpy(version, start, MINIMUM(sizeof(version), (size_t)(stop - start + 1)));
+         break;
+      }
+   }
+   fclose(fp);
+
+   snprintf(buffer, bufferLen, "macOS %s", version[0] ? version : "");
+}
+
+void Platform_getRelease(char** string) {
+   *string = Generic_unameRelease(Platform_getOSRelease);
```

Comment:
I believe the callback can be an overkill here.

How about:

```suggestion
   static char osRelease[100] = {0};
   if (!osRelease[0])
      Platform_getOSRelease(osRelease, sizeof(osRelease));
   Generic_unameRelease(osRelease);
```

Yes, this way you don't need a callback function, and you can locally cache the `osRelease` string just parsed from the `.plist` file.

Reply:
I think the callback is OK and it helps keep the logic consistent across the major platforms.